misc/docerina/src/main/java/org/ballerinalang/docgen/Generator.java

```diff
@@ -485,13 +485,38 @@ private static BClass getClassModel(ClassDefinitionNode classDefinitionNode, Sem
         if (containsToken(classDefinitionNode.classTypeQualifiers(), SyntaxKind.CLIENT_KEYWORD)) {
             return new Client(name, description, isDeprecated, fields, functions, isReadOnly, isIsolated, isService);
         } else if (containsToken(classDefinitionNode.classTypeQualifiers(), SyntaxKind.LISTENER_KEYWORD)
-                || name.equals("Listener")) {
+                || isListenerModel(functions)) {
             return new Listener(name, description, isDeprecated, fields, functions, isReadOnly, isIsolated, isService);
         } else {
             return new BClass(name, description, isDeprecated, fields, functions, isReadOnly, isIsolated, isService);
         }
     }
 
+    private static boolean isListenerModel(List<Function> classFunctions) {
+        boolean isStartIncluded = false;
+        boolean isAttachIncluded = false;
+        boolean isDetachIncluded = false;
+        boolean isGracefulStopIncluded = false;
+        boolean isImmediateStopIncluded = false;
+
+        for (Function function : classFunctions) {
+            if (function.name.equals("'start")) {
+                isStartIncluded = true;
+            } else if (function.name.equals("attach")) {
+                isAttachIncluded = true;
+            } else if (function.name.equals("detach")) {
+                isDetachIncluded = true;
+            } else if (function.name.equals("gracefulStop")) {
+                isGracefulStopIncluded = true;
+            } else if (function.name.equals("immediateStop")) {
```

Comment:
@NipunaMadhushan Also just wondering is it enough to check the name or do we need to validate the signature too